common/persistence/shardManager_test.go

```diff
@@ -0,0 +1,399 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package persistence
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/types"
+)
+
+func TestShardManagerGetName(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	store := NewMockShardStore(ctrl)
+	store.EXPECT().GetName().Return("name")
+
+	manager := NewShardManager(store)
+
+	assert.Equal(t, "name", manager.GetName())
+}
+
+func TestShardManagerClose(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	store := NewMockShardStore(ctrl)
+	store.EXPECT().Close().Times(1)
+
+	manager := NewShardManager(store)
+	manager.Close()
+}
+
+func TestShardManagerCreateShard(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	tests := map[string]struct {
+		request          *CreateShardRequest
+		internalRequest  *InternalCreateShardRequest
+		serializer       PayloadSerializer
+		internalResponse error
+		expected         error
+	}{
+		"Success": {
+			request: &CreateShardRequest{
+				ShardInfo: sampleShardInfo(),
+			},
+			internalRequest: &InternalCreateShardRequest{
+				ShardInfo: sampleInternalShardInfo(t),
+			},
+			internalResponse: nil,
+			expected:         nil,
+		},
+		"Serialization Failure": {
+			request: &CreateShardRequest{
+				ShardInfo: sampleShardInfo(),
+			},
+			serializer: failingSerializer(ctrl),
+			expected:   fmt.Errorf("serialization"),
+		},
+		"Error Response": {
+			request: &CreateShardRequest{
+				ShardInfo: sampleShardInfo(),
+			},
+			internalRequest: &InternalCreateShardRequest{
+				ShardInfo: sampleInternalShardInfo(t),
+			},
+			internalResponse: fmt.Errorf("error"),
+			expected:         fmt.Errorf("error"),
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			store := NewMockShardStore(ctrl)
+			if test.internalRequest != nil {
+				store.EXPECT().CreateShard(gomock.Any(), gomock.Eq(test.internalRequest)).Return(test.internalResponse)
+			}
+
+			manager := NewShardManager(store)
+			if test.serializer != nil {
+				manager.(*shardManager).serializer = test.serializer
```

Comment:
a better alternative pattern for this kind of overrides is to use `WithX` options. [Example](https://github.com/uber/cadence/blob/d7479caa65df2064e0f162f9c9b726c27e5d0d23/common/cache/domainCache.go#L146)